File: services/asset-worker/pysrc/ingest.py
#!/usr/bin/env python3
"""model_ingest geometry step — normalize a source model and emit a LOD glb + metadata.

Run as a SUBPROCESS by the Rust worker (ADR-007: never in-process, for crash isolation + retry),
NOT imported. It is the CPU half of the asset pipeline (architecture.md §5.3): trimesh loads the source
(.glb/.stl/.obj/.3mf), recenters it to the origin, measures its bounding box (the dims that prefill the
Product), and exports a glb the storefront's model-viewer loads. The heavier gltf-transform LOD pass
(Draco/meshopt/KTX2 compression) is a later slice; this produces a plain, correct glb first.

Contract with the Rust wrapper (src/ingest.rs):
  argv: <input-model-path> <output-dir>
  stdout: ONE line of JSON — {"dimsMm":[w,d,h], "glbPath":"...", "triangles":N, "watertight":bool,
          "objectNames":[...], "structuredGlbPath":"..."?}
          objectNames = f-2 part-mapping options ([] for a single-mesh source); structuredGlbPath = the f-4
          named-objects glb, OMITTED when the source has no named objects (the viewer falls back to glbPath).
  exit 0 + that manifest on success; exit non-zero + a stderr reason on failure (a bad/unsupported
  model is a permanent failure the wrapper reports as `failed`; the wrapper classifies by exit path).
"""

import json
import os
import shutil
import signal
import subprocess
import sys
from pathlib import Path

import trimesh

# Wall-clock budget for the whole ingest (trimesh runs IN this process, so a hang here would stall the
# concurrency=1 worker forever). Set by ingest.rs from worker config (INGEST_TIMEOUT_SECS); default matches.
TIMEOUT_SECS = int(os.environ.get("INGEST_TIMEOUT_SECS", "300"))
# sysexits EX_TEMPFAIL — the timeout exit the Rust wrapper classifies as TRANSIENT (redeliver),
# unlike any other non-zero exit (permanent). Keep in sync with ingest.rs/render.rs EXIT_TIMEOUT.
EXIT_TIMEOUT = 75
# Triangle cap for EVERY served glb (fused + structured) — the anti-download measure: the browser only
# ever receives a decimated display shell; the printable source never leaves the bucket. 50k = the
# asset-worker rule's LOD ceiling. Decimation failure on an over-cap mesh FAILS the ingest (fail-closed:
# never fall back to shipping the full-res mesh).
MAX_TRIS = int(os.environ.get("INGEST_MAX_TRIS", "50000"))


def decimated(mesh, target: int):
    """Quadric-decimate to ≤ target faces; a mesh already under cap passes through untouched."""
    if mesh.faces.shape[0] <= target:
        return mesh
    return mesh.simplify_quadric_decimation(face_count=target)


def compress(glb_path: Path) -> None:
    """Draco-compress a served glb in place via gltf-transform (Node CLI baked in the worker image).
    Pure size optimization — the decimation cap above is the security measure — so best-effort: a
    missing CLI (local dev) or a compressor quirk leaves the uncompressed glb and logs to stderr.
    `draco` (not `optimize`) on purpose: optimize's weld/prune/join can rename or merge the named
    objects/materials the f-2..f-5 chain keys on. model-viewer decodes KHR_draco_mesh_compression
    with its default gstatic decoder (model-3d-viewer.tsx)."""
    cli = shutil.which("gltf-transform")
    if cli is None:
        print("gltf-transform not found — serving uncompressed glb", file=sys.stderr)
        return
    tmp = glb_path.with_suffix(".draco.glb")
    try:
        subprocess.run([cli, "draco", str(glb_path), str(tmp)], check=True, capture_output=True)
        tmp.replace(glb_path)
    except Exception as e:  # noqa: BLE001 — best-effort; _IngestTimeout is a BaseException, never swallowed here
        detail = getattr(e, "stderr", b"") or b""
        print(f"draco compress failed ({e}): {detail.decode(errors='replace')[:500]}", file=sys.stderr)
        tmp.unlink(missing_ok=True)


def structured_artifact(input_path: str, translation, out_dir: str):
    """f-2/f-4: load the SOURCE as a Scene (names + materials intact — the fused export below drops them),
    apply the SAME recenter translation as the fused export (D-B: identical vector, else the ADR-038 camera
    pose de-frames one derivative), and emit (object-name list, structured glb Path | None). Best-effort: any
    load/export quirk → ([], None), so neither the dropdown options nor the structured glb can fail an
    otherwise-good ingest. A single-mesh source (an STL) has no named objects → ([], None). Names are in MODEL
    order (deterministic per file), matching the whole f-2/f-4 chain."""
    try:
        scene = trimesh.load(input_path)
        geometry = getattr(scene, "geometry", None)  # a Scene has .geometry; a lone Trimesh (STL) does not
        if not geometry:
            return [], None
        names = list(geometry.keys())
        # f-3: name each object's material after the object, so the storefront live viewer can address it via
        # model-viewer `getMaterialByName(objectName)` (model-viewer recolors by MATERIAL — mesh→material is
        # not public API). A geometry with no material (e.g. vertex-colour visuals) is skipped → that part
        # simply won't recolour (graceful), never a crash.
        total = sum(g.faces.shape[0] for g in geometry.values()) or 1
        for name, geom in list(geometry.items()):
            material = getattr(getattr(geom, "visual", None), "material", None)
            if material is not None:
                material.name = name
            # Decimate each part to its proportional share of the cap (floor 64 keeps tiny parts intact).
            # Decimation returns a bare mesh → reattach the named material so f-3 recolouring still works.
            share = max(int(MAX_TRIS * geom.faces.shape[0] / total), 64)
            if geom.faces.shape[0] > share:
                slim = geom.simplify_quadric_decimation(face_count=share)
                if material is not None:
                    slim.visual = trimesh.visual.TextureVisuals(material=material)
                scene.geometry[name] = slim
        scene.apply_translation(translation)  # the SAME vector as the fused export → the two stay aligned
        structured_path = Path(out_dir) / "model_structured.glb"
        structured_path.write_bytes(trimesh.exchange.gltf.export_glb(scene))
        compress(structured_path)
        return names, structured_path
    except Exception:  # noqa: BLE001 — the structured artifact is optional metadata; never fail the ingest over it
        return [], None


def ingest(input_path: str, out_dir: str) -> dict:
    # force='mesh' collapses a multi-part scene into one mesh so extents/export are well-defined.
    mesh = trimesh.load(input_path, force="mesh")
    if mesh.is_empty or mesh.vertices.shape[0] == 0:
        raise ValueError("model has no geometry")

    # Normalize: recenter the bounding-box center to the origin (the worker owns placement; the source may be
    # modelled anywhere). Deterministic and lossless — no scaling, so real-world mm dims survive. The SAME
    # translation is reused for the structured export (D-B) so both derivatives frame identically.
    translation = -mesh.bounding_box.centroid
    mesh.apply_translation(translation)

    w, d, h = (round(float(x), 3) for x in mesh.extents)  # bounding-box dims, model units (mm)

    out = Path(out_dir)
    out.mkdir(parents=True, exist_ok=True)
    glb_path = out / "model.glb"
    # Dims/triangles/watertight above are measured on the ORIGINAL; only the SERVED artifact is decimated.
    glb_path.write_bytes(trimesh.exchange.gltf.export_glb(trimesh.Scene(decimated(mesh, MAX_TRIS))))
    compress(glb_path)

    # ponytail: a SECOND load (in structured_artifact) reads names + exports the structured glb — keeps THIS
    # fused path (what ADR-038's pose is curated against) byte-identical; dedupe to one load only if latency bites.
    names, structured_path = structured_artifact(input_path, translation, out_dir)

    manifest = {
        "dimsMm": [w, d, h],
        "glbPath": str(glb_path),
        "triangles": int(mesh.faces.shape[0]),
        "watertight": bool(mesh.is_watertight),
        "objectNames": names,  # f-2: named parts for the editor mapping (empty for STL)
    }
    if structured_path is not None:
        manifest["structuredGlbPath"] = str(structured_path)  # f-4: named-objects glb (omitted when none)
    return manifest


class _IngestTimeout(BaseException):
    """Raised by the SIGALRM handler. BaseException on purpose: the blanket `except Exception`
    fallbacks (main, structured_artifact) must never swallow the timeout into a permanent failure."""


def main() -> int:
    if len(sys.argv) != 3:
        print("usage: ingest.py <input-model-path> <output-dir>", file=sys.stderr)
        return 2
    def _on_alarm(_signum, _frame):
        raise _IngestTimeout()

    signal.signal(signal.SIGALRM, _on_alarm)
    signal.alarm(TIMEOUT_SECS)
    try:
        manifest = ingest(sys.argv[1], sys.argv[2])
    except _IngestTimeout:
        print(f"ingest timed out after {TIMEOUT_SECS}s", file=sys.stderr)
        return EXIT_TIMEOUT
    except Exception as e:  # noqa: BLE001 — any load/export failure is a permanent, reportable error
        print(f"ingest failed: {e}", file=sys.stderr)
        return 1
    print(json.dumps(manifest))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: services/asset-worker/src/ingest.rs
//! The `model_ingest` geometry step — a thin Rust wrapper around `pysrc/ingest.py` (trimesh), run as a
//! SUBPROCESS (ADR-007: crash-isolation + retry, never in-process). It turns a source model file into a
//! recentered glb + its bounding-box dims. The S3 fetch that feeds it and the upload that consumes its
//! glb — plus wiring this into the `model_ingest` `Processor` — are the next slice; THIS ships the
//! verified transform and its output parsing.

use std::path::Path;
use std::process::Command;

use serde::Deserialize;

use crate::processor::ProcessError;

/// sysexits EX_TEMPFAIL — the DISTINCT exit both python scripts return on a wall-clock timeout
/// (ingest.py SIGALRM / render.py subprocess timeout). Mapped to **Transient** (redeliver): a hang is an
/// environment fault (a CUDA stall, a wedged load), not the model's — unlike other non-zero exits.
pub const EXIT_TIMEOUT: i32 = 75;

/// The manifest `ingest.py` prints on success (one line of JSON, camelCase). `dims_mm` is the recentered
/// bounding box `[w, d, h]` in model units (mm) — the values that prefill Product; `glb_path` is the
/// exported glb the caller uploads to lumin-assets.
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Manifest {
    pub dims_mm: [f64; 3],
    pub glb_path: String,
    pub triangles: u64,
    pub watertight: bool,
    /// The object/material names in the SOURCE model (f-2). The recentered export fuses to one mesh, so
    /// these are read separately (ingest.py `scene.geometry.keys()`) and never affect dims/glb. Empty for a
    /// single-mesh source (an STL). `#[serde(default)]` keeps an older ingest.py that omits the field parseable.
    #[serde(default)]
    pub object_names: Vec<String>,
    /// f-4: path to the STRUCTURED glb (named objects/materials preserved, same recenter as `glb_path`) the
    /// caller uploads as a second derivative. Absent for a single-mesh source / an export quirk — the manifest
    /// omits it and the viewer falls back to the fused glb. `#[serde(default)]` → None when the field is absent.
    #[serde(default)]
    pub structured_glb_path: Option<String>,
}

/// run_ingest invokes `python ingest.py <input> <out_dir>` and parses the manifest. Error classification
/// drives the WorkQueue lifecycle (processor::ProcessError):
///   - non-zero exit → **Permanent** (the script's contract: a bad/unsupported model — retrying can't
///     help; the worker reports `failed`).
///   - spawn failure (python/script missing) or an exit-0-but-unparseable stdout → **Transient** (an
///     environment/tooling fault, not the model's — redeliver, so a mis-provisioned image never burns a
///     good job as `failed`).
///   - exit `EXIT_TIMEOUT` (75) → **Transient**: the script killed itself on its wall-clock budget
///     (INGEST_TIMEOUT_SECS, worker config) — a hang is an environment fault, redeliver.
pub fn run_ingest(
    python: &str,
    script: &Path,
    input: &Path,
    out_dir: &Path,
    timeout_secs: u64,
) -> Result<Manifest, ProcessError> {
    let output = Command::new(python)
        .arg(script)
        .arg(input)
        .arg(out_dir)
        .env("INGEST_TIMEOUT_SECS", timeout_secs.to_string())
        .output()
        .map_err(|e| ProcessError::Transient(format!("spawn {python}: {e}")))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        if output.status.code() == Some(EXIT_TIMEOUT) {
            return Err(ProcessError::Transient(format!(
                "ingest timed out after {timeout_secs}s: {}",
                stderr.trim()
            )));
        }
        let code = output
            .status
            .code()
            .map(|c| c.to_string())
            .unwrap_or_else(|| "signal".into());
        return Err(ProcessError::Permanent(format!(
            "ingest exited {code}: {}",
            stderr.trim()
        )));
    }
    parse_manifest(&output.stdout)
}

/// parse_manifest reads the script's stdout JSON. A tool that exits 0 but emits junk is a script/env bug,
/// not the model's fault → Transient (redeliver) rather than marking a good model `failed`.
fn parse_manifest(stdout: &[u8]) -> Result<Manifest, ProcessError> {
    serde_json::from_slice(stdout)
        .map_err(|e| ProcessError::Transient(format!("unparseable ingest manifest: {e}")))
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    fn crate_dir() -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
    }

    // --- pure parsing (CI-portable, no subprocess) ---
    #[test]
    fn parses_the_scripts_manifest() {
        let m = parse_manifest(
            br#"{"dimsMm":[20.0,30.0,40.0],"glbPath":"/out/model.glb","triangles":12,"watertight":true}"#,
        )
        .expect("parse manifest");
        assert_eq!(m.dims_mm, [20.0, 30.0, 40.0]);
        assert_eq!(m.triangles, 12);
        assert!(m.watertight);
        assert!(m.glb_path.ends_with(".glb"));
    }

    #[test]
    fn parses_optional_object_names_and_structured_path() {
        // f-2 objectNames + f-4 structuredGlbPath flow through when present…
        let m = parse_manifest(
            r#"{"dimsMm":[1.0,2.0,3.0],"glbPath":"/o/m.glb","triangles":4,"watertight":false,"objectNames":["Chao đèn","Đế"],"structuredGlbPath":"/o/model_structured.glb"}"#
                .as_bytes(),
        )
        .expect("parse with names + structured");
        assert_eq!(
            m.object_names,
            vec!["Chao đèn".to_string(), "Đế".to_string()]
        );
        assert_eq!(
            m.structured_glb_path.as_deref(),
            Some("/o/model_structured.glb")
        );
        // …and an older ingest.py that omits BOTH still parses (serde default → empty/None), never a fail.
        let m0 = parse_manifest(
            br#"{"dimsMm":[1.0,2.0,3.0],"glbPath":"/o/m.glb","triangles":4,"watertight":false}"#,
        )
        .expect("parse without optionals");
        assert!(m0.object_names.is_empty());
        assert!(m0.structured_glb_path.is_none());
    }

    #[test]
    fn exit0_but_junk_stdout_is_transient() {
        // A script that exits 0 with unparseable output is an env/script bug, not a bad model → retry.
        assert!(matches!(
            parse_manifest(b"not json"),
            Err(ProcessError::Transient(_))
        ));
    }

    // --- subprocess classification (CI-portable: no trimesh needed) ---
    #[test]
    fn missing_interpreter_is_transient() {
        let err = run_ingest(
            "definitely-not-a-real-interpreter-xyz",
            Path::new("ingest.py"),
            Path::new("in.obj"),
            Path::new("/tmp/out"),
            300,
        )
        .unwrap_err();
        assert!(matches!(err, ProcessError::Transient(_)), "got {err:?}");
    }

    #[test]
    fn nonzero_exit_is_permanent() {
        // `sh /nonexistent-script …` exits non-zero — stands in for the ingest script rejecting a bad
        // model (its contract: non-zero exit = permanent). Uses sh (present in CI), not trimesh.
        let err = run_ingest(
            "sh",
            Path::new("/nonexistent-ingest-script-xyz"),
            Path::new("in.obj"),
            Path::new("/tmp/out"),
            300,
        )
        .unwrap_err();
        assert!(matches!(err, ProcessError::Permanent(_)), "got {err:?}");
    }

    #[test]
    fn exit_timeout_code_is_transient() {
        // A script exiting EXIT_TIMEOUT (its wall-clock budget fired) must redeliver, not burn the job.
        let dir = std::env::temp_dir().join("lumin-ingest-timeout-test");
        std::fs::create_dir_all(&dir).unwrap();
        let script = dir.join("exit75.sh");
        std::fs::write(&script, "exit 75\n").unwrap();
        let err =
            run_ingest("sh", &script, Path::new("in.obj"), Path::new("/tmp/out"), 1).unwrap_err();
        assert!(matches!(err, ProcessError::Transient(_)), "got {err:?}");
    }

    // --- the REAL transform, gated on a trimesh-capable python (INGEST_PYTHON). Skips in CI (no
    // trimesh), runs locally / on the box. Proves the script+wrapper against the committed fixture. ---
    #[test]
    fn real_trimesh_ingests_the_fixture() {
        let Ok(python) = std::env::var("INGEST_PYTHON") else {
            eprintln!("skip: set INGEST_PYTHON to a python with trimesh to run the real ingest");
            return;
        };
        let script = crate_dir().join("pysrc/ingest.py");
        let input = crate_dir().join("testdata/box.obj");
        let out = std::env::temp_dir().join("lumin-ingest-test");
        let m = run_ingest(&python, &script, &input, &out, 300).expect("real ingest");
        // The fixture is a 20×30×40 box translated off-origin — recentering must not change the dims.
        assert_eq!(m.dims_mm, [20.0, 30.0, 40.0]);
        assert!(std::path::Path::new(&m.glb_path).exists(), "glb written");
        assert!(m.triangles > 0);
    }
}
